---
title: "MAGMA Celltyping"
author: "Nathan Skene"
date: "`r Sys.Date()`"
output: rmarkdown::github_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Setup MAGMA

## Set parameters to be used for the analysis

```{r }
# MAGMA parameters
kb_upstream = 10
kb_downstream = 1.5
gwasN = 0 # Set this equal to the sample size of the GWAS

# MAGMA file paths
gene_loc_dir = sprintf("%s/data/",system.file(package="MAGMA.Celltyping"))
gene_loc_files = c("NCBI37.3.gene.loc")
if(!sum(gene_loc_files %in% list.files(gene_loc_dir))==length(gene_loc_files)){stop("ERROR: not all gene_loc_files are located in gene_loc_dir")}
genome_ref_dir = "/Users/ns9/OneDrive - University College London/MAGMA/g1000_eur/g1000_eur"
```

# Prepare cell type specificity data

## Install and load all the required R packages and data

The EWCE package comes with a celltype specificity dataset which we use as an example.  The One2One package is used to obtain 1:1 orthologs.

```{r }
# Load the celltype data
data(ctd)

# Load the mouse to human 1:1 orthologs
data(ortholog_data_Mouse_Human)
```

## Prepare quantile groups for celltypes

First we need to calculate the quantile groups for each celltype within the single cell dataset. This is done using the `prepare.quantile.groups` function. If your single cell dataset is not from mouse, then change the specificity_species argument. If you wish to use a smaller number of bins then 

```{r }
ctd = prepare.quantile.groups(ctd,specificity_species="mouse",numberOfBins=40)

# Examine how the quantile groups look
print(ctd[[1]]$quantiles[c("Gfap","Dlg4","Aif1"),])
print(table(ctd[[1]]$quantiles[,1]))
```

# Run MAGMA

## Download summary statistics file & check it is properly formatted

We need to have a summary statistics file to analyse as input. As an example download that for risk taking, based on the UK Biobank, generated by Ben Neale's group.

The function `check.sumstats.formatted.for.magma` does some basic processing to get it into the right format.

```{r }
library(R.utils)
gwas_sumstats_path = "~/Downloads/202040.assoc.tsv"
if(!file.exists(gwas_sumstats_path)){
    download.file("https://www.dropbox.com/s/27yrfcsngenzab1/2040.assoc.tsv.gz?raw=1",destfile=gwas_sumstats_path)
    gunzip(sprintf("%s.gz",gwas_sumstats_path),gwas_sumstats_path)
}
col_headers = check.sumstats.formatted.for.magma(gwas_sumstats_path)


library(rsnps)
source("https://bioconductor.org/biocLite.R")
library("SNPlocs.Hsapiens.dbSNP144.GRCh37")
#library(SNPlocs.Hsapiens.dbSNP144.GRCh38)
snps <- SNPlocs.Hsapiens.dbSNP144.GRCh37


biocLite("SNPlocs.Hsapiens.dbSNP144.GRCh37") # [The SNPs in this package are mapped to reference genome GRCh37.p13]


# Note, use the names of the SNPs printed out from the header above to find which genome build this was mapped with
# -- use dbSNP to check this: https://www.ncbi.nlm.nih.gov/projects/SNP/
gene_loc_dir = "/Users/ns9/Workshop"
genomeLocFile=sprintf("%s/NCBI37.3.gene.loc",gene_loc_dir)
#
```

## Map SNPs to Genes

```{r }
magma_cmd = sprintf("%smagma --annotate window=%s,%s --snp-loc '%s' --gene-loc '%s' --out '%s'",magma_path,kb_upstream,kb_downstream,gwas_sumstats_path,genomeLocFile,gwas_sumstats_path)
system(magma_cmd)

if("N" %in% column_headers){n_arg = "ncol=N"}else{n_arg = sprintf("N=%s",gwasN)}
# SCHIZ CLOZUK N=35802

magma_cmd = sprintf("%smagma --bfile '%s' --pval '%s' %s --gene-annot '%s.genes.annot' --out '%s'",magma_path,genome_ref_dir,gwas_sumstats_path,n_arg,gwas_sumstats_path,gwas_sumstats_path)
magma_cmd
system(magma_cmd)


# window size: 10 kb upstream and 1.5 kb downstream of each gene 
# model: snpwise-unweighted
# option --gene-covar onesided
# p-values need correcting to one sided
# For the conditional analysis, we used the condition modifier of the —gene-covar parameter to condition on each of the significant cell types
```

## Run the analysis

```{r }
# First write the quantile data to a file
##- Need to match quantiles to the genes in the genes.out file
genesOut = read.table(sprintf("%s.genes.out",gwas_sumstats_path),stringsAsFactors = FALSE)
data(all_hgnc_wtEntrez)
#load("/Users/ns9/Box Sync/MAGMA.Celltyping/data/all_hgnc_wtEntrez.rda")
colnames(all_hgnc_wtEntrez)[1] = "human.symbol"
data(ortholog_data_Mouse_Human)
entrez_mgi = merge(all_hgnc_wtEntrez,ortholog_data_Mouse_Human$orthologs_one2one[,2:3],by="human.symbol")
entrez_mgi = entrez_mgi[!is.na(entrez_mgi$entrezgene),]

entrez_mgi = entrez_mgi[entrez_mgi$mouse.symbol %in% rownames(ctd[[annotLevel]]$quantiles),]
annotLevel=1
quantDat = ctd[[annotLevel]]$quantiles[entrez_mgi$mouse.symbol,]
quantDat2 = data.frame(entrez=entrez_mgi$entrezgene,quantDat)
quantDat2 = quantDat2[!duplicated(quantDat2$entrez),]
geneCovarFile="/Users/ns9/geneCovarMagma.csv"
#write.csv(quantDat2,file=geneCovarFile,quote=FALSE,row.names=FALSE)
write.table(quantDat2,file=geneCovarFile,quote=FALSE,row.names=FALSE,sep="\t")

analysis_name = "test"
magma_cmd = sprintf("%smagma --gene-results '%s.genes.raw' --gene-covar '%s' --out '%s.%s'",magma_path,gwas_sumstats_path,geneCovarFile,gwas_sumstats_path,analysis_name)
print(magma_cmd)
system(magma_cmd)
```

## Now rerun with another CTD file

```{r }
ctd=list()
load("/Users/ns9/Box Sync/Patrick & Julien/celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_level1_thresh0_trim0.rda")
ctd[[1]] = list()
ctd[[1]]$specificity = celltype_data[[1]]$cell_dists
ctd[[1]]$mean_exp = celltype_data[[1]]$all_scts
ctd[[1]]$annot = celltype_data[[1]]$annot
load("/Users/ns9/Box Sync/Patrick & Julien/celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_level2_thresh0_trim0.rda")
ctd[[2]] = list()
ctd[[2]]$specificity = celltype_data[[1]]$cell_dists
ctd[[2]]$mean_exp = celltype_data[[1]]$all_scts
ctd[[2]]$annot = celltype_data[[1]]$annot
save(ctd,file="/Users/ns9/Box Sync/Patrick & Julien/celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0.rda")


# ff="ctd_StenWholeBrain10xGenomics2018_StringentGeneFiltering.rda"
#ff="ctd_StenWholeBrain10xGenomics2018_RlyExtremeGeneFiltering.rda"
#path="~/Box Sync/MAGMA.Celltyping/"
ff="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0.rda"
path="/Users/ns9/Box Sync/Patrick & Julien//"
load(sprintf("%s%s",path,ff))
ctd = prepare.quantile.groups(ctd,"mouse","human",40)
colnames(ctd[[1]]$quantiles) = gsub("\\(|\\)|;","",gsub(" ","",colnames(ctd[[1]]$quantiles)))
colnames(ctd[[2]]$quantiles) = gsub("\\(|\\)|;","",gsub(" ","",colnames(ctd[[2]]$quantiles)))

load("/Users/ns9/Box Sync/MAGMA.Celltyping/data/all_hgnc_wtEntrez.rda")
colnames(all_hgnc_wtEntrez)[1] = "human.symbol"
entrez_mgi = merge(all_hgnc_wtEntrez,ortholog_data_Mouse_Human$orthologs_one2one[,2:3],by="human.symbol")
entrez_mgi = entrez_mgi[!is.na(entrez_mgi$entrezgene),]

entrez_mgi = entrez_mgi[entrez_mgi$mouse.symbol %in% rownames(ctd[[1]]$quantiles),]
for(annotLevel in 1:length(ctd)){
    quantDat = ctd[[annotLevel]]$quantiles[entrez_mgi$mouse.symbol,]
    quantDat2 = data.frame(entrez=entrez_mgi$entrezgene,quantDat)
    quantDat2 = quantDat2[!duplicated(quantDat2$entrez),]
    colnames(quantDat2)[2:dim(quantDat2)[2]]
    geneCovarFile="/Users/ns9/geneCovarMagma.csv"
    #write.csv(quantDat2,file=geneCovarFile,quote=FALSE,row.names=FALSE)
    write.table(quantDat2,file=geneCovarFile,quote=TRUE,row.names=FALSE,sep="\t")
    
    analysis_name = ff #"ctd_StenWholeBrain10xGenomics2018_ExtremeGeneFilterin"
    magma_cmd = sprintf("%smagma --gene-results '%s.genes.raw' --gene-covar '%s' onesided --out '%s.%s.annotLevel%s'",magma_path,gwas_sumstats_path,geneCovarFile,gwas_sumstats_path,analysis_name,annotLevel)
    print(magma_cmd)
    system(magma_cmd)
}    

library(ggplot2)
analysis_name=ff#"ctd_StenWholeBrain10xGenomics2018_ExtremeGeneFilterin"
baseline_res = list()
for(annotLevel in 1:length(ctd)){
    res = read.table(file=sprintf("%s.%s.annotLevel%s.gcov.out",gwas_sumstats_path,analysis_name,annotLevel),stringsAsFactors = FALSE)
    colnames(res) = res[1,]
    res = res[-1,]
    res$P = as.numeric(res$P)
    res = res[order(res$P,decreasing = TRUE),]
    res$COVAR = factor(res$COVAR,levels=res$COVAR)
    pdf(file=sprintf("%s.%s.annotLevel%s.gcov.pdf",gwas_sumstats_path,analysis_name,annotLevel),width=10,height=1+(dim(res)[1]/10))
    print(ggplot(data=res)+geom_bar(aes(y=P,x=COVAR),stat="identity")+scale_y_log10()+coord_flip())
    dev.off()
    res$CONTROL = "BASELINE"
    res$CONTROL_label = "BASELINE"
    res$ANNOTLEVEL=annotLevel
    baseline_res[[annotLevel]] = res
}


# Conditional analysis

library(ggplot2)
annotLevel=1
res = read.table(file=sprintf("%s.%s.annotLevel%s.gcov.out",gwas_sumstats_path,analysis_name,annotLevel),stringsAsFactors = FALSE)
colnames(res) = res[1,]
res = res[-1,]
res$P = as.numeric(res$P)
res = res[order(res$P,decreasing = TRUE),]
res$Q = p.adjust(res$P,method="BH")
res = res[order(res$P),]
signifCells = res[res$Q<0.05,"COVAR"]
if(length(signifCells)>6){signifCells=res[res$Q<0.05,"COVAR"][1:6]}
count=allRes=0
for(controlFor in signifCells){
    quantDat = ctd[[annotLevel]]$quantiles[entrez_mgi$mouse.symbol,]
    quantDat2 = data.frame(entrez=entrez_mgi$entrezgene,quantDat)
    quantDat2 = quantDat2[!duplicated(quantDat2$entrez),]
    geneCovarFile="/Users/ns9/geneCovarMagma.csv"
    #write.csv(quantDat2,file=geneCovarFile,quote=FALSE,row.names=FALSE)
    write.table(quantDat2,file=geneCovarFile,quote=TRUE,row.names=FALSE,sep="\t")
    
    analysis_name = sprintf("ControlFor_%s",controlFor)
    magma_cmd = sprintf("%smagma --gene-results '%s.genes.raw' --gene-covar '%s' onesided condition=\\\"%s\\\" --out '%s.%s.annotLevel%s'",magma_path,gwas_sumstats_path,geneCovarFile,controlFor,gwas_sumstats_path,analysis_name,annotLevel)
    print(magma_cmd)
    system(magma_cmd)    
    
    cond_res = read.table(file=sprintf("%s.%s.annotLevel%s.gcov.out",gwas_sumstats_path,analysis_name,annotLevel),stringsAsFactors = FALSE)
    colnames(cond_res) = cond_res[1,]
    cond_res = cond_res[-1,]
    cond_res$P = as.numeric(cond_res$P)
    cond_res = cond_res[order(cond_res$P,decreasing = TRUE),]
    cond_res$CONTROL = controlFor
    count = count + 1
    if(count==1){
        allRes = cond_res
    }else{
        allRes = rbind(allRes,cond_res)
    }
}
allRes$ANNOTLEVEL = annotLevel
allRes_base = allRes
allRes$CONTROL_label = ""
allRes = rbind(allRes,baseline_res[[1]])
fullRes = rbind(allRes,baseline_res[[2]])
save(fullRes,file=sprintf("%s.%s.annotLevel%s.allResults.Rda",gwas_sumstats_path,analysis_name,annotLevel))
write.csv(fullRes,file=sprintf("%s.%s.annotLevel%s.allResults.csv",gwas_sumstats_path,analysis_name,annotLevel))
allRes = allRes[allRes$COVAR %in% signifCells,]
source("R/get.written.cellnames.r")
allRes$COVAR = get.written.cellnames(allRes$COVAR,file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")
allRes$CONTROL[allRes$CONTROL!="BASELINE"] = get.written.cellnames(allRes$CONTROL[allRes$CONTROL!="BASELINE"],file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")
allRes$CONTROL_label[allRes$CONTROL!="BASELINE"] = sprintf("Conditioned on %s",allRes$CONTROL[allRes$CONTROL!="BASELINE"])
subRes = allRes[allRes$COVAR %in% signifCells,]

res$COVAR = factor(res$COVAR,levels=res$COVAR)
analysis_name = sprintf("%s-ConditionalFacets",ff)
library(cowplot)
pdf(file=sprintf("%s.%s.annotLevel%s.gcov.pdf",gwas_sumstats_path,analysis_name,annotLevel),width=14,height=1+2*(dim(res)[1]/10))
outPlot = ggplot(data=allRes)+geom_bar(aes(y=P,x=COVAR,fill=COVAR),stat="identity")+scale_y_log10()+coord_flip()+facet_wrap(~CONTROL_label)+
        ylab(expression(-log[10](pvalue))) + xlab("") + theme(legend.position="none")+
        geom_hline(yintercept=as.numeric(0.05/dim(allRes)[1]),colour="black")
print(outPlot)
dev.off()






# 
# ldsc=read.csv("/Users/ns9/Desktop/Edinburgh/LDSC_GWAS_Results_stepSize_0.1.csv")
# ldsc = ldsc[ldsc$file %in% c("celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_level1_thresh0_trim0","celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_level2_thresh0_trim0"),]
# ldsc = ldsc[ldsc$percentile==1.0,]
# ldsc = ldsc[,-1]
# ldsc = ldsc[order(ldsc$Enrichment_p),]
# ldsc = ldsc[!is.na(ldsc$celltype),]
# ldsc$Enrichment_P_oneTailed = ldsc$Enrichment_p
# ldsc$Enrichment_P_oneTailed[ldsc$Enrichment>1] = ldsc$Enrichment_p[ldsc$Enrichment>1]/2
# ldsc$Enrichment_P_oneTailed[ldsc$Enrichment<1] = 1-(ldsc$Enrichment_p[ldsc$Enrichment<1]/2)
# ldsc = ldsc[order(ldsc$Enrichment_P_oneTailed),]
# write.csv(ldsc,file="LDSC_Results_for_DavidHill.csv")
# 
# ct_names = ldsc_res[,c("celltype","celltype2")]
# write.csv(unique(ct_names),file="ldsc_ct_names.csv")
# 
# ctd_names_1 = colnames(ctd[[1]]$quantiles)
# ctd_names_fixed_1 = get.written.cellnames(colnames(ctd[[1]]$quantiles),file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")
# ctd_names_2 = colnames(ctd[[2]]$quantiles)
# ctd_names_fixed_2 = get.written.cellnames(colnames(ctd[[2]]$quantiles),file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")
# ctd_names_dt = data.frame(ctd_names=c(ctd_names_1,ctd_names_2),writtenNames=c(ctd_names_fixed_1,ctd_names_fixed_2),stringsAsFactors = FALSE)
# ctd_names_dt_uniq = unique(ctd_names_dt[order(ctd_names_dt$writtenNames),])
# write.csv(ctd_names_dt_uniq,file="ctd_names.csv",row.names = FALSE)
# ctd_uniq_names = ctd_names_dt_uniq$writtenNames
# sort(ctd_uniq_names[!ctd_uniq_names %in% ldsc_res$celltype2])
# sort(ldsc_res$celltype2[!ldsc_res$celltype2 %in% ctd_uniq_names])
# 
# original_ctd_names = gsub("\\(|\\)|;","",gsub(" ","",c(colnames(ctd[[1]]$quantiles),colnames(ctd[[2]]$quantiles))))
# sort(original_ctd_names[!original_ctd_names %in% ldsc_res$celltype])
# sort(ldsc_res$celltype[!ldsc_res$celltype %in% original_ctd_names])
# 
# original_ctd_names = gsub("\\(|\\)|;","",gsub("\\.","",magma_res$COVAR))
# sort(original_ctd_names[!original_ctd_names %in% ldsc_res$celltype])
# sort(ldsc_res$celltype[!ldsc_res$celltype %in% original_ctd_names])

```

Merge MAGMA results with LDSC results

```{r }
edu_magma = cbind(gwas="Education_trait_1.txt",read.csv("/Users/ns9/Desktop/Edinburgh/Education_trait_1.txt.ControlFor_SerotonergicNeuron.annotLevel1.allResults.csv",stringsAsFactors = FALSE)[,-1])
hi_magma  = cbind(gwas="HI.txt",read.csv("/Users/ns9/Desktop/Edinburgh/HI.txt.snploc.ControlFor_interneurons.annotLevel1.allResults.csv",stringsAsFactors = FALSE)[,-1])
magma_res = rbind(edu_magma,hi_magma)
ldsc_res  = read.csv("/Users/ns9/Desktop/Edinburgh/LDSC_Results_for_DavidHill.csv",stringsAsFactors = FALSE)[,-1]
ldsc_res$ANNOTLEVEL = 1
ldsc_res$ANNOTLEVEL[grep("level2",ldsc_res$file)] = 2
ldsc_res = ldsc_res[,c("gwas","celltype","ANNOTLEVEL","Enrichment_oneTailed")]

# The MAGMA names lose all symbols, i.e. +, - or ;... drop those from LDSC and use to match the two together
ldsc_res$celltype2 = gsub("\\-|\\+|;","\\.",ldsc_res$celltype)
cellname_mapping = unique(ldsc_res[,c("celltype","celltype2")])
rownames(cellname_mapping) = as.character(cellname_mapping$celltype2)
magma_res$COVAR = as.character(cellname_mapping[magma_res$COVAR,]$celltype)
magma_res$CONTROL[magma_res$CONTROL!="BASELINE"] = as.character(cellname_mapping[magma_res$CONTROL[magma_res$CONTROL!="BASELINE"],]$celltype)
ldsc_res = ldsc_res[,c("gwas","celltype","ANNOTLEVEL","Enrichment_oneTailed")]
ldsc_res$CONTROL="BASELINE"
magma_res = magma_res[,c("gwas","COVAR","ANNOTLEVEL","P","CONTROL")]
colnames(ldsc_res) = c("gwas","celltype","ANNOTLEVEL","P_Enrichment_oneTailed","ControlledFor")
colnames(magma_res) = c("gwas","celltype","ANNOTLEVEL","P_Enrichment_oneTailed","ControlledFor")

# Join MAGMA and LDSC together
both_res = rbind(cbind(method="LDSC",ldsc_res),cbind(method="MAGMA",magma_res))
both_res$q = p.adjust(both_res$P_Enrichment_oneTailed,method="BH")
both_res$writtenName = get.written.cellnames(both_res$celltype,file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")

# Amend the names of the GWAS studies
both_res$gwas[both_res$gwas=="Education_trait_1.txt"] = "Educational Achievement"
both_res$gwas[both_res$gwas=="HI.txt"] = "Household Income"

# Plot joint results
library(cowplot)
for(aL in unique(both_res$ANNOTLEVEL)){
    subDat = both_res[both_res$ControlledFor=="BASELINE" & both_res$ANNOTLEVEL==aL,]
    subDat = subDat[order(subDat$P_Enrichment_oneTailed,decreasing=TRUE),]
    subDat$writtenName = factor(subDat$writtenName,levels=unique(subDat$writtenName))
    pdf(file=sprintf("Edu+HI.JointResults.annotLevel%s.pdf",aL),width=14,height=1+2*(length(unique(subDat$writtenName))/10))
    outPlot = ggplot(data=subDat)+geom_bar(aes(y=P_Enrichment_oneTailed,x=writtenName,fill=method),position="dodge",stat="identity")+scale_y_log10()+coord_flip()+
            ylab(expression(-log[10](pvalue))) + xlab("") + facet_wrap(~gwas) +
            geom_hline(yintercept=as.numeric(0.05/dim(both_res)[1]),colour="black")
    #outPlot
    print(outPlot)
    dev.off()
}


for(gwas in unique(both_res$gwas)){
    
    subDat = both_res[both_res$ControlledFor!="BASELINE" & both_res$gwas==gwas,]
    subDat$ControlledFor = get.written.cellnames(subDat$ControlledFor,file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")
    pdf(file=sprintf("Edu+HI.%s.JointResults.Conditional.gcov.pdf",gwas),width=14,height=1+(dim(subDat)[1]/13))
    outPlot = ggplot(data=subDat)+geom_bar(aes(y=P_Enrichment_oneTailed,x=writtenName,fill=writtenName),stat="identity")+scale_y_log10()+coord_flip()+facet_wrap(~ControlledFor)+
            ylab(expression(-log[10](pvalue))) + xlab("") + theme(legend.position="none")+
            geom_hline(yintercept=as.numeric(0.05/dim(both_res)[1]),colour="black")
    print(outPlot)
    dev.off()
}

save(both_res,file="/Users/ns9/Desktop/Edinburgh/To Send/both_res.rda")
```


```{r }
#unique(magma_res$COVAR)[!unique(magma_res$COVAR) %in% ldsc_res$celltype]

source("/Users/ns9/Box Sync/MAGMA.Celltyping/R/get.written.cellnames.r")
magma_res$COVAR2 = get.written.cellnames(magma_res$COVAR,file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")
ldsc_res$celltype2 = get.written.cellnames(ldsc_res2$celltype,file="celltype_data_allKImouse_wtHypo_MergedStriatal_1to1only_thresh0_trim0")
sort(unique(magma_res$COVAR[magma_res$ANNOTLEVEL==1]))
sort(unique(magma_res$COVAR2[magma_res$ANNOTLEVEL==1]))
sort(unique(magma_res$COVAR2[magma_res$ANNOTLEVEL==2]))
sort(unique(ldsc_res$celltype[ldsc_res$ANNOTLEVEL==1]))


unique(magma_res$COVAR)[!unique(magma_res$COVAR) %in% ldsc_res$celltype]
unique(magma_res$COVAR)[!unique(magma_res$COVAR) %in% ldsc_res$celltype]
unique(ldsc_res$celltype[!ldsc_res$celltype %in% magma_res$COVAR])
```




