---
title: "MAGMA_Celltyping: Getting started"
author: "Brian M. Schilder"
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  BiocStyle::html_document:
vignette: >
    %\VignetteIndexEntry{MAGMA_Celltyping} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---


```{r setup, include=TRUE, message=FALSE}
library(MAGMA.Celltyping)
library(ggplot2)
library(dplyr)
```


# Setup 

Specify where you want the large files to be downloaded to. 

*NOTE*: Make sure you change `storage_dir` to somewhere other than `tempdir()` 
if you want to make sure the results aren't deleted after this R session closes!

```{r, eval=FALSE}
storage_dir <- tempdir()
```


# Prepare data

## Import GWAS

- We need to have a summary statistics file to analyse as input.  
- As an example, you can download UK Biobank summary statistics 
for 'fluid_intelligence' using `get_example_gwas()`.

Here we provide a 
pre-munged version of the above file.

### Munging 

Our lab have created [`MungeSumstats`](https://github.com/neurogenomics/MungeSumstats), a robust Bioconductor package for formatting multiple types
of summary statistics files. We highly recommend processing your GWAS summary statistics with
`MungeSumstats` before continuing.
See the *full_workflow* vignette for more details.

The minimum info needed *after* munging is:  
- "SNP", "CHR", and "BP" as first three columns.
- It has at least one of these columns: "Z","OR","BETA","LOG_ODDS","SIGNED_SUMSTAT"

```{r, eval=FALSE}
path_formatted <- MAGMA.Celltyping::get_example_gwas(
  trait = "prospective_memory",
  munged = TRUE)
```

## Map SNPs to Genes

Note you can input the genome build of your summary statistics for this step or
it can be inferred if left `NULL`:

```{r, eval=FALSE}
genesOutPath <- MAGMA.Celltyping::map_snps_to_genes(
  path_formatted = path_formatted,
  genome_build = "GRCh37")
```


## CellTypeDataset

### Import CTD 

`ewceData` provides a number of CellTypeDatasets (CTD) to be used a cell-type
transcriptomic signature reference files.

If you want to create your own single-cell transcriptomic reference, you'll need
to first convert it to CTD using the instructions found in the `EWCE` package
[documentation here](https://github.com/NathanSkene/EWCE/).

```{r, eval=FALSE } 
ctd <- ewceData::ctd()
```

Note that the cell type dataset loaded in the code above is the Karolinksa cortex/hippocampus data only. For the full Karolinska dataset with hypothalamus and midbrain instead use the following:

```
ctd <- MAGMA.Celltyping::get_ctd("ctd_allKI")
```

Or for the DRONC seq or AIBS datasets use:

```
ctd <- get_ctd("ctd_Tasic")
ctd <- get_ctd("ctd_DivSeq")
ctd <- get_ctd("ctd_AIBS")
ctd <- get_ctd("ctd_DRONC_human")
ctd <- get_ctd("ctd_DRONC_mouse")
ctd <- get_ctd("ctd_BlueLake2018_FrontalCortexOnly")
ctd <- get_ctd("ctd_BlueLake2018_VisualCortexOnly")
ctd <- get_ctd("ctd_Saunders")
```
 
### Run cell-type enrichment analyses

```{r}
MAGMA_results <- MAGMA.Celltyping::celltype_associations_pipeline(
  magma_dirs = dirname(genesOutPath),
  ctd = ctd, 
  sctSpecies = "mouse", 
  ctd_name = "Zeisel2015", 
  run_linear = TRUE, 
  run_top10 = TRUE,
  run_condition = FALSE)

```

# Plot results

```{r}
MAGMA.Celltyping::gather_results(MAGMA_results = MAGMA_results)
```


# Session Info

<details>

```{r}
utils::sessionInfo()
```

</details>
<hr>
