---
title: "OpenGWAS"
author: "Brian M. Schilder"
date: "`r Sys.Date()`"
csl: nature.csl
output:
  BiocStyle::html_document:
vignette: >
    %\VignetteIndexEntry{OpenGWAS} 
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

```{r style, echo=FALSE, results='asis', message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE, message  = FALSE)
```

```{r setup, include=TRUE}
#### To install the dev branch of MAGMA.Celltyping, use:
# remotes::install_github("NathanSkene/MAGMA_Celltyping@bschilder_dev")

# devtools::install_github("neurogenomics/MungeSumstats")
library(MAGMA.Celltyping)
library(MungeSumstats)

library(ggplot2)
library(dplyr)

save_dir <- tempdir()
# save_dir <- "/Volumes/bms20/projects/neurogenomics-lab/live/GWAS_sumstats/OpenGWAS"
```

# Munge sum stats

## Find sum stats

```{r} 
metagwas <- MungeSumstats::find_sumstats(traits = c("Alzheimer","Parkinson",
                                                    "amyotrophic","dementia"))
knitr::kable(metagwas)
```

## Import sum stats

```{r}
gwas <- MungeSumstats::import_sumstats(ids = metagwas$id, 
                                       save_dir = save_dir, 
                                       nThread = 10, 
                                       parallel_across_ids = FALSE,
                                       force_new = FALSE) 
```


# Create metadata file

```{r}
munged_files <- list.files(save_dir,".tsv.gz", full.names = TRUE)
munged_ids <- gsub(".tsv.gz","",basename(munged_files))
metagwas_all <- MungeSumstats::find_sumstats(ids = munged_ids)
metagwas_all$path <- munged_files
```

## Check genome builds

`MungeSumstats` automatically keeps log files in the `tempdir()`. But if you restart your R session mid-way, they'll be erased.

Instead, infer the genome build here.

```{r}
builds <- MungeSumstats::get_genome_builds(sumstats_list = munged_files, 
                                           header_only = 1000, 
                                           names_from_paths =TRUE)
table(metagwas_all$build_inferred)
```

## Filter metadata

Must have sample size.

```{r}
metagwas_all <- metagwas_all %>% 
  dplyr::mutate(N = ifelse(is.na(sample_size), sum(ncase, ncontrol, na.rm = TRUE), sample_size))
metagwas_all <- subset(metagwas_all, !is.na(N))
```


## Save metadata

```{r} 
metagwas_all <- metagwas_all %>% `rownames<-`(gsub("-|[:]|[.]","_",rownames(metagwas_all)))
utils::write.csv(metagwas_all, file.path(save_dir,"OpenGWAS_metadata.csv"))
```


# Map SNPs to genes

## Read metadata

```{r}
metagwas_all <- read.csv(file.path(save_dir,"OpenGWAS_metadata.csv"), row.names = 1)

metagwas_all$trait_group <- stringr::str_extract(
  tolower(metagwas_all$trait),
  "(alzheimer)|(parkinson)|(amyotrophic)|(frontotemporal dementia)|(dementia)")
metagwas_all[is.na(metagwas_all$trait_group),"trait_group"] <- "other"

```

 
## Map SNPs

```{r}
magma_files <- parallel::mclapply(1:nrow(metagwas_all), function(i){
  MAGMA.Celltyping:::message_parallel("----- ", metagwas_all$id[i]," -----")
  MAGMA.Celltyping::map_snps_to_genes(path_formatted = metagwas_all$path[i],
                                      genome_build = metagwas_all$build_inferred[i],
                                      upstream_kb = 35,  
                                      downstream_kb = 10,
                                      N = metagwas_all$N[i],
                                      genome_ref_path = genome_ref_path,
                                      force_new = FALSE)
}, mc.cores = 10) %>% `names<-`(metagwas_all$id)
```


# Construct gene x trait matrix

Can construct gene x trait matrix from gene-level MAGMA results across multiple GWAS traits.

```{r}
magma_matrix <- phenomix::gene_trait_matrix(magma_files = magma_files,
                                            metric = "ADJ_ZSTAT", 
                                            save_path = file.path(save_dir,"OpenGWAS_neurodegenGWAS_ADJZSTAT.rds"),
                                            mc.cores = 10) 
```


# Session Info

<details>

```{r}
utils::sessionInfo()
```

</details>
