#' Import precomputed CellTypeDatasets
#'
#' Import CellTypeDatasets (CTD) generated by
#' \href{https://github.com/NathanSkene/EWCE}{EWCE}.
#'
#' \itemize{
#'
#' \item{"ctd_AIBS"}{CTD file with AIBS human cortex data.
#' \href{http://www.hjerling-leffler-lab.org/data/scz_singlecell/}{Reference}.}
#'
#' \item{"ctd_allKI"}{CTD file with cortex, hippocampus,
#' hypothalamus and midbrain.
#' \href{http://www.hjerling-leffler-lab.org/data/scz_singlecell/}{Reference}.}
#'
#' \item{"ctd_BlueLake2018_FrontalCortexOnly"}{
#' CTD file with data from humans.}
#'
#' \item{"ctd_BlueLake2018_VisualCortexOnly"}{
#' CTD file with data from humans.}
#'
#' \item{"ctd_Saunders"}{
#' CTD file with data from humans.}
#'
#' \item{"ctd_DescartesHuman"}{
#' CTD file with scRNA-seq data from human embryo across multiple organ systems.
#' \href{https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/}{Reference}.
#' }
#'
#' \item{"ctd_DRONC_human"}{
#' CTD file with DRONC-seq data from humans.
#' \href{http://www.hjerling-leffler-lab.org/data/scz_singlecell/}{Reference}.
#' }
#'
#' \item{"ctd_DRONC_mouse"}{
#' CTD file with Tasic data from mouse.
#' \href{http://www.hjerling-leffler-lab.org/data/scz_singlecell/}{Reference}.
#' }
#'
#' \item{"ctd_DivSeq"}{
#' CTD file with DivSeq data.
#' \href{http://www.hjerling-leffler-lab.org/data/scz_singlecell/}{Reference}.
#' }
#'
#' \item{"ctd_Tasic"}{
#' CTD file with Tasic data.
#' \href{http://www.hjerling-leffler-lab.org/data/scz_singlecell/}{Reference}.
#' }
#'
#' \item{"ctd_Zeisel2018"}{
#' CTD file with data from mice.
#' \href{http://www.hjerling-leffler-lab.org/data/scz_singlecell/}{Reference}.
#' }
#'
#' }
#'
#' @source
#' \code{
#' ## Move from data/ to Releases on GitHub to speed up MAGMA.Celltyping
#' all_ctd <- list.files("data/","^ctd*", full.names = TRUE)
#' # all_ctd2 <- list.files(
#' #     file.path("../../model_celltype_conservation/processed_data/",
#' #               "EWCE/standardized_CTD"),
#' #     "*.rds$", full.names = TRUE)
#' all_ctd=c(all_ctd, 
#'           file.path("../../model_celltype_conservation/processed_data",
#'                     "EWCE/standardized_CTD/DescartesHuman.rds"))
#' for(save_path in all_ctd){
#'     piggyback::pb_upload(file = save_path,
#'                          repo = "neurogenomics/MAGMA_Celltyping",
#'                          overwrite = TRUE)
#' }
#' 
#' arg_list <- rlang::fn_fmls(fn = MAGMA.Celltyping::get_ctd)
#' ctd_names <- eval(arg_list$ctd_name)
#' for(d in ctd_names){
#'     message(d)
#'     ctd <- MAGMA.Celltyping::get_ctd(ctd_name = d)
#'     for(lv in seq_len(length(ctd))){
#'         if(!"annot" %in% names(ctd[[lv]])){
#'             ctd[[lv]]$annot <- data.frame(celltype=colnames(ctd[[lv]]$mean_exp))
#'         } 
#'     }
#'     # ctd <- EWCE::standardise_ctd(ctd = ctd, input_species = "human")
#'     names(ctd) <- paste0("level",seq_len(length(ctd)))
#'     save_path <- file.path(tempdir(),paste0(d,".rda"))
#'     save(ctd, file = save_path) 
#'     piggyback::pb_upload(file = save_path,
#'                          repo = "neurogenomics/MAGMA_Celltyping",
#'                          overwrite = TRUE)
#' }



#' }
#'
#' @param ctd_name Name of the CellTypeDatset to import.
#' @param storage_dir Folder in which to save the CellTypeDatset.
#' @return CellTypeDataset object
#'
#' @examples
#' ctd <- MAGMA.Celltyping::get_ctd("ctd_AIBS")
#' @export
get_ctd <- function(ctd_name = c(
                        "ctd_AIBS",
                        "ctd_allKI",
                        "ctd_BlueLake2018_FrontalCortexOnly",
                        "ctd_BlueLake2018_VisualCortexOnly",
                        "ctd_DescartesHuman",
                        "ctd_DivSeq",
                        "ctd_DRONC_human",
                        "ctd_DRONC_mouse",
                        "ctd_Saunders",
                        "ctd_Tasic",
                        "ctd_Zeisel2018"
                    ),
                    storage_dir = tempdir()) {
    tmp <- get_data(
        fname = paste0(ctd_name, ".rda"),
        storage_dir = storage_dir
    )
    obj <- load_rdata(tmp)
    return(obj)
}
